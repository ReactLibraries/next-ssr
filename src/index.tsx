import React, {
  ReactNode,
  useContext,
  useId,
  useRef,
  useCallback,
  useSyncExternalStore,
  createContext,
} from 'react';

const DATA_NAME = '__NEXT_DATA_PROMISE__';

type StateType<T> = { data?: T; error?: unknown; isLoading: boolean; fetcher: () => Promise<T> };
type Render = () => void;
type ContextType = {
  values: { [key: string]: StateType<unknown> };
  promises: Promise<unknown>[];
  finished: boolean;
  renderMap: Map<string | number, Set<Render>>;
};

/**
 * Context for asynchronous data management
 */
const promiseContext = createContext<ContextType>(undefined as never);

/**
 * Rendering event propagation
 */
const render = (renderMap: Map<string | number, Set<Render>>, key: string | number) =>
  renderMap.get(key)?.forEach((render) => render());

/**
 * Asynchronous data loading
 */
const loader = <T,>(key: string | number, context: ContextType, fetcher?: () => Promise<T>) => {
  const { promises, values, renderMap } = context;
  const _fetcher = fetcher ?? values[key]?.fetcher;
  if (!_fetcher) throw new Error('Empty by fetcher');
  const value = { data: values[key]?.data, error: undefined, isLoading: true, fetcher: _fetcher };
  values[key] = value;
  render(renderMap, key);
  const promise = _fetcher();
  if (typeof window === 'undefined') {
    promises.push(promise);
  }
  promise
    .then((v) => {
      values[key] = { data: v, error: undefined, isLoading: false, fetcher: _fetcher };
      render(renderMap, key);
    })
    .catch((error) => {
      values[key] = { data: undefined, error, isLoading: false, fetcher: _fetcher };
      render(renderMap, key);
    });
  return promise;
};

/**
 * hook for re-loading
 */
export const useReload = (key: string | number) => {
  const context = useContext(promiseContext);
  return useCallback(() => {
    loader(key, context);
  }, [context, key]);
};

/**
 * Asynchronous data acquisition hook for SSR
 */
export const useSSR = <T,>(
  fetcher: () => Promise<T>,
  { key }: { key?: string | number } = {}
): StateType<T> & { reload: () => void } => {
  const context = useContext(promiseContext);
  const { values, renderMap } = context;
  const id = useId();
  const cacheKey = key ?? id;

  const value = useSyncExternalStore(
    (callback) => {
      const renderSet = renderMap.get(cacheKey) ?? new Set<Render>();
      renderMap.set(cacheKey, renderSet);
      renderSet.add(callback);
      return () => renderSet.delete(callback);
    },
    () => values[cacheKey] as StateType<T>,
    () => values[cacheKey] as StateType<T>
  );

  const reload = useCallback(() => {
    return loader(cacheKey, context, fetcher);
  }, [cacheKey, context, fetcher]);
  if (!value) {
    const promise = reload();
    if (typeof window === 'undefined') {
      throw promise;
    }
  } else if (!value.fetcher) {
    value.fetcher = fetcher;
  }

  return { ...value, reload };
};

/**
 * Transfer of SSR data to clients
 */
const DataRender = () => {
  const context = useContext(promiseContext);
  if (typeof window === 'undefined' && !context.finished)
    throw Promise.allSettled(context.promises).then((v) => {
      context.finished = true;
      return v;
    });
  return (
    <script
      id={DATA_NAME}
      type="application/json"
      dangerouslySetInnerHTML={{ __html: JSON.stringify(context.values) }}
    />
  );
};

/**
 * Context data initialisation hook
 */
const useContextValue = () => {
  const refContext = useRef<ContextType>({
    values: {},
    promises: [],
    finished: false,
    renderMap: new Map<string, Set<Render>>(),
  });
  if (typeof window !== 'undefined' && !refContext.current.finished) {
    const node = document.getElementById(DATA_NAME);
    if (node) refContext.current.values = JSON.parse(node.innerHTML);
    refContext.current.finished = true;
  }
  return refContext.current;
};

/**
 * Provider for asynchronous data management
 */
export const SSRProvider = ({ children }: { children: ReactNode }) => {
  const value = useContextValue();
  return (
    <promiseContext.Provider value={value}>
      {children}
      <DataRender />
    </promiseContext.Provider>
  );
};
