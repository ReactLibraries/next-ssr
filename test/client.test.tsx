/**
 * @jest-environment jest-environment-jsdom
 */

import React, { useEffect } from 'react';
import { createRoot } from 'react-dom/client';
import { act } from 'react-dom/test-utils';
import { SSRProvider, useReload, useSSR } from '../src';

let prev;
beforeEach(() => {
  prev = global.IS_REACT_ACT_ENVIRONMENT;
  global.IS_REACT_ACT_ENVIRONMENT = true;
});

afterEach(() => {
  global.IS_REACT_ACT_ENVIRONMENT = prev;
});

const Component01 = () => {
  const { data } = useSSR<number>(async () => {
    return 100;
  });
  return <div id="data1">{data}</div>;
};

const Component02 = () => {
  const { data } = useSSR<number>(
    async () => {
      return data ? data + 100 : 200;
    },
    { key: 'Component02' }
  );
  const refetch = useReload('Component02');
  useEffect(() => {
    refetch();
  }, [refetch]);
  return <div id="data2">{data}</div>;
};

const Component03 = () => {
  const { error } = useSSR<number>(
    async () => {
      throw 'error';
    },
    { key: 'Component03' }
  );
  return <div id="data3">{String(error)}</div>;
};

const Component04 = () => {
  const refetch = useReload('error');
  useEffect(() => {
    try {
      refetch();
    } catch (_) {}
  }, [refetch]);
  return <div id="data4">data4</div>;
};

it('Client', async () => {
  const container = document.createElement('div');
  container.innerHTML = `<div id="data1">100</div><div id="data2">200</div><div id="data3">error</div><script id="__NEXT_DATA_PROMISE__" type="application/json">{":R5:":{"data":100,"isLoading":false},"Component02":{"data":200,"isLoading":false},"Component03":{"error":"error","isLoading":false}}</script>`;
  document.body.appendChild(container);
  const root = createRoot(container);
  await act(async () => {
    root.render(
      <SSRProvider>
        <Component01 />
        <Component02 />
        <Component03 />
        <Component04 />
      </SSRProvider>
    );
  });
  expect(container.childNodes).toMatchSnapshot();
  await act(() => {
    root.unmount();
  });
  container.remove();
});
