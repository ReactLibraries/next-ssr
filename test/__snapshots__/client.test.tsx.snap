// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client 1`] = `
NodeList [
  <div
    id="data1"
  >
    100
  </div>,
  <div
    id="data2"
  >
    300
  </div>,
  <div
    id="data3"
  >
    error
  </div>,
  <div
    id="data4"
  >
    data4
  </div>,
  <script
    id="__NEXT_DATA_PROMISE__"
    type="application/json"
  >
    {":R5:":{"data":100,"isLoading":false},"Component02":{"data":200,"isLoading":false},"Component03":{"error":"error","isLoading":false},":r0:":{"isLoading":true}}
  </script>,
]
`;
